runtime/src/main/java/com/dylibso/chicory/runtime/ByteArrayMemory.java

```diff
@@ -0,0 +1,350 @@
+        this.dataSegments = dataSegments;
+        if (dataSegments == null) {
+            return;
+        }
+
+        for (var s : dataSegments) {
+            if (s instanceof ActiveDataSegment) {
+                var segment = (ActiveDataSegment) s;
+                var offsetExpr = segment.offsetInstructions();
+                var data = segment.data();
+                var offset = (int) computeConstantValue(instance, offsetExpr)[0];
+                checkBounds(
+                        offset,
+                        data.length,
+                        sizeInBytes(),
+                        (msg) -> new UninstantiableException(msg));
+                System.arraycopy(data, 0, buffer, offset, data.length);
+            } else if (s instanceof PassiveDataSegment) {
+                // Passive segment should be skipped
+            } else {
+                throw new ChicoryException("Data segment should be active or passive: " + s);
+            }
+        }
+    }
+
+    private static void checkBounds(
+            int addr, int size, int limit, Function<String, ChicoryException> exceptionFactory) {
+        if (addr < 0 || size < 0 || addr > limit || (size > 0 && ((addr + size) > limit))) {
+            var errorMsg =
+                    "out of bounds memory access: attempted to access address: "
+                            + addr
+                            + " but limit is: "
+                            + limit
+                            + " and size: "
+                            + size;
+            throw exceptionFactory.apply(errorMsg);
+        }
+    }
+
+    private static RuntimeException outOfBoundsException(
+            RuntimeException e, int addr, int size, int limit) {
+        if (e instanceof IndexOutOfBoundsException
+                || e instanceof BufferOverflowException
+                || e instanceof BufferUnderflowException
+                || e instanceof IllegalArgumentException
+                || e instanceof NegativeArraySizeException) {
+            var errorMsg =
+                    "out of bounds memory access: attempted to access address: "
+                            + addr
+                            + " but limit is: "
+                            + limit
+                            + " and size: "
+                            + size;
+            return new WasmRuntimeException(errorMsg);
+        } else {
+            return e;
+        }
+    }
+
+    @Override
+    public void initPassiveSegment(int segmentId, int dest, int offset, int size) {
+        var segment = dataSegments[segmentId];
+        write(dest, segment.data(), offset, size);
+    }
+
+    private int sizeInBytes() {
+        return PAGE_SIZE * nPages;
+    }
+
+    @Override
+    public void write(int addr, byte[] data, int offset, int size) {
+        try {
+            System.arraycopy(data, offset, buffer, addr, size);
+        } catch (RuntimeException e) {
```

Comment:
Apparently, the JVM is able to optimize bound checks, thus is convenient to use `try` - `catch` than explicit checks.

We are catching `RuntimeException` to keep the method body short and easily inlinable.

Reply:
can you add this note as a comment (even in the heading of the class is fine)